lib/perfdata/elasticsearchdatastreamwriter.cpp

```diff
@@ -0,0 +1,878 @@
+		[this](const Checkable::Ptr& checkable, const CheckResult::Ptr& cr, const MessageOrigin::Ptr&) {
+			CheckResultHandler(checkable, cr);
+		}
+	);
+}
+
+/* Pause is equivalent to Stop, but with HA capabilities to resume at runtime. */
+void ElasticsearchDatastreamWriter::Pause()
+{
+	m_HandleCheckResults.disconnect();
+	m_FlushTimer->Stop(true);
+	m_Paused = true;
+	m_WorkQueue.Enqueue([this]() { Flush(); });
+	m_WorkQueue.Join();
+
+	Log(LogInformation, "ElasticsearchDatastreamWriter")
+		<< "'" << GetName() << "' paused.";
+
+	ObjectImpl<ElasticsearchDatastreamWriter>::Pause();
+}
+
+void ElasticsearchDatastreamWriter::ManageIndexTemplate()
+{
+	AssertOnWorkQueue();
+
+	String templatePath = ICINGA_PKGDATADIR "/elasticsearch/index-template.json";
+	std::ifstream templateFile{ templatePath, std::ifstream::in };
+	if (!templateFile.is_open()) {
+		Log(LogCritical, "ElasticsearchDatastreamWriter")
+			<< "Could not open index template file: " << templatePath;
+		return;
+	}
+	std::ostringstream templateStream;
+	templateStream << templateFile.rdbuf();
+	String templateJson = templateStream.str();
+	templateFile.close();
+	Log(LogDebug, "ElasticsearchDatastreamWriter")
+		<< "Read index template from " << templatePath;
+	Log(LogDebug, "ElasticsearchDatastreamWriter") << templateJson;
+
+	Url::Ptr indexTemplateUrl = new Url();
+	indexTemplateUrl->SetScheme(GetEnableTls() ? "https" : "http");
+	indexTemplateUrl->SetHost(GetHost());
+	indexTemplateUrl->SetPort(GetPort());
+	indexTemplateUrl->SetPath({ "_index_template", "icinga2-metrics" });
+
+	Url::Ptr componentTemplateUrl = new Url();
+	componentTemplateUrl->SetScheme(GetEnableTls() ? "https" : "http");
+	componentTemplateUrl->SetHost(GetHost());
+	componentTemplateUrl->SetPort(GetPort());
+	componentTemplateUrl->SetPath({ "_component_template", "metrics-icinga2@custom" });
+	componentTemplateUrl->SetQuery({ {"create", "true"} });
+
+	while (true) {
+		if (m_Paused) {
+			Log(LogInformation, "ElasticsearchDatastreamWriter")
+				<< "Shutdown in progress, aborting index template management.";
+			return;
+		}
+
+		try {
+			Dictionary::Ptr jsonResponse = TrySend(componentTemplateUrl, "{\"template\":{}}");
+			Log(LogInformation, "ElasticsearchDatastreamWriter")
+				<< "Successfully installed component template 'icinga2@custom'.";
+		} catch (const StatusCodeException& es) {
+			if (es.GetStatusCode() == http::status::bad_request) {
+				Log(LogInformation, "ElasticsearchDatastreamWriter")
+					<< "Component template 'metrics-icinga2@custom' already exists, skipping creation.";
+				// Continue to install/update index template
+			} else {
+				Log(LogWarning, "ElasticsearchDatastreamWriter")
+					<< "Failed to install component template 'icinga2@custom', retrying in 5 seconds: " << DiagnosticInformation(es, false);
+				Log(LogDebug, "ElasticsearchDatastreamWriter")
+					<< "Additional information:\n" << DiagnosticInformation(es, true);
```

Comment:
I think, all occurrences of this such log (with Debug level) can be dropped entirely, as you already have the other log line with relevant info. You shouldn't need any detailed stack traces here but just the error message itself received from Elasticsearch.